File: src/dolphin_context_actions/converters/ffmpeg_tools.py
"""Pick an ffmpeg binary that actually has the encoder a preset asks for.

ffmpeg builds ship different encoder sets. A Homebrew/linuxbrew build commonly
omits libvorbis while the distro build in /usr/bin has it, so hardcoding
"ffmpeg" made "Convert to OGG" fail with a raw "Unknown encoder" dump.
"""

import shutil
import subprocess

# Searched in order. The PATH binary wins when it can do the job.
_CANDIDATE_BINARIES = ("ffmpeg", "/usr/bin/ffmpeg", "/usr/local/bin/ffmpeg")

_encoder_cache: dict[str, frozenset[str]] = {}


def _encoders(binary: str) -> frozenset[str]:
    """Encoder names the given ffmpeg build reports."""
    if binary in _encoder_cache:
        return _encoder_cache[binary]

    names: set[str] = set()
    try:
        result = subprocess.run(
            [binary, "-hide_banner", "-encoders"],
            capture_output=True, text=True, timeout=15,
        )
        for line in result.stdout.splitlines():
            # Rows look like " A....D libmp3lame  libmp3lame MP3 ...". The
            # legend above them ("V..... = Video", "A..... = Audio", ...) has
            # the same 6-char flag-column width, so parts[1] == "=" is the
            # extra check that tells a legend line from a real encoder row.
            parts = line.split()
            if len(parts) >= 2 and len(parts[0]) == 6 and parts[1] != "=" and not line.startswith(" -"):
                names.add(parts[1])
    except Exception:
        pass

    resolved = frozenset(names)
    _encoder_cache[binary] = resolved
    return resolved


def resolve(*encoders: str) -> str | None:
    """Return an ffmpeg path providing every encoder, or None if none does."""
    wanted = {e for e in encoders if e}
    for binary in _CANDIDATE_BINARIES:
        path = shutil.which(binary)
        if not path:
            continue
        if wanted <= _encoders(path):
            return path
    return None


def missing_encoder_message(label: str, *encoders: str) -> str:
    """Explain which encoder is unavailable and how to get it."""
    names = ", ".join(f"<b>{e}</b>" for e in encoders if e)
    return (
        f"Cannot convert to {label}.\n\n"
        f"No ffmpeg on this system provides: {names}\n\n"
        "Install an ffmpeg build that includes it:\n"
        "• <tt>sudo dnf install ffmpeg</tt> (Fedora)\n"
        "• <tt>sudo apt install ffmpeg</tt> (Debian/Ubuntu/Mint)\n"
        "• <tt>sudo pacman -S ffmpeg</tt> (Arch/Manjaro)"
    )


def clear_cache():
    """Forget probed encoder sets. Used by tests."""
    _encoder_cache.clear()

File: src/converters/ffmpeg_tools.rs
use std::collections::HashMap;
use std::process::Command;
use std::sync::Mutex;

use std::sync::OnceLock;

static CANDIDATE_BINARIES: &[&str] = &["ffmpeg", "/usr/bin/ffmpeg", "/usr/local/bin/ffmpeg"];

fn encoder_cache() -> &'static Mutex<HashMap<String, Vec<String>>> {
    static CACHE: OnceLock<Mutex<HashMap<String, Vec<String>>>> = OnceLock::new();
    CACHE.get_or_init(|| Mutex::new(HashMap::new()))
}

fn encoders(binary: &str) -> Vec<String> {
    if let Ok(cache) = encoder_cache().lock() {
        if let Some(names) = cache.get(binary) {
            return names.clone();
        }
    }

    let mut names = Vec::new();
    if let Ok(output) = Command::new(binary)
        .args(["-hide_banner", "-encoders"])
        .output()
    {
        for line in String::from_utf8_lossy(&output.stdout).lines() {
            let parts: Vec<&str> = line.split_whitespace().collect();
            if parts.len() >= 2
                && parts[0].len() == 6
                && parts[1] != "="
                && !line.starts_with(" -")
            {
                names.push(parts[1].to_string());
            }
        }
    }

    if let Ok(mut cache) = encoder_cache().lock() {
        cache.insert(binary.to_string(), names.clone());
    }
    names
}

pub fn resolve(wanted: &[&str]) -> Option<String> {
    let wanted: Vec<&str> = wanted.iter().copied().filter(|e| !e.is_empty()).collect();
    for binary in CANDIDATE_BINARIES {
        let path = which::which(binary).ok()?;
        let path_s = path.display().to_string();
        let have = encoders(&path_s);
        if wanted.iter().all(|enc| have.iter().any(|h| h == enc)) {
            return Some(path_s);
        }
    }
    None
}

pub fn missing_encoder_message(label: &str, encoders: &[&str]) -> String {
    let names = encoders
        .iter()
        .filter(|e| !e.is_empty())
        .map(|e| format!("<b>{e}</b>"))
        .collect::<Vec<_>>()
        .join(", ");
    format!(
        "Cannot convert to {label}.\n\n\
         No ffmpeg on this system provides: {names}\n\n\
         Install an ffmpeg build that includes it:\n\
         • <tt>sudo dnf install ffmpeg</tt> (Fedora)\n\
         • <tt>sudo apt install ffmpeg</tt> (Debian/Ubuntu/Mint)\n\
         • <tt>sudo pacman -S ffmpeg</tt> (Arch/Manjaro)"
    )
}

pub fn clear_cache() {
    if let Ok(mut cache) = encoder_cache().lock() {
        cache.clear();
    }
}
